src/lotus/types.rs

```diff
@@ -0,0 +1,196 @@
+    #[allow(dead_code)]
+    pub balance: String,
+    #[allow(dead_code)]
+    pub code: CIDMap,
+    #[allow(dead_code)]
+    pub state: State,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+pub struct Receipt {
+    #[allow(dead_code)]
+    exit_code: u32,
+    #[allow(dead_code)]
+    r#return: String,
+    #[allow(dead_code)]
+    gas_used: u64,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "PascalCase")]
+pub struct MpoolPushMessageResponse {
+    pub message: MpoolPushMessageInner,
+    #[serde(rename = "CID")]
+    pub cid: CIDMap,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "PascalCase")]
+pub struct MpoolPushMessageInner {
+    to: String,
+    from: String,
+    pub value: String,
+    pub method: MethodNum,
+    pub params: String,
+
+    pub nonce: u64,
+    #[serde(rename = "GasLimit")]
+    pub gas_limit: u64,
+    #[serde(rename = "GasFeeCap")]
+    pub gas_fee_cap: String,
+    #[serde(rename = "GasPremium")]
+    pub gas_premium: String,
+    pub version: u16,
+
+    #[serde(rename = "CID")]
+    pub cid: CIDMap,
+}
+
+impl MpoolPushMessageInner {
+    pub fn get_root_cid(&self) -> Option<Cid> {
+        self.cid
+            .cid
+            .as_ref()
+            .map(|s| Cid::from_str(s).expect("server sent invalid cid"))
+    }
+
+    pub fn to(&self) -> anyhow::Result<Address> {
+        Ok(Address::from_str(&self.to)?)
+    }
+
+    pub fn from(&self) -> anyhow::Result<Address> {
+        Ok(Address::from_str(&self.from)?)
+    }
+}
+
+pub struct MpoolPushMessage {
+    pub to: Address,
+    pub from: Option<Address>,
+    pub value: TokenAmount,
+    pub method: MethodNum,
+    pub params: Vec<u8>,
+
+    pub nonce: Option<u64>,
+    pub gas_limit: Option<TokenAmount>,
+    pub gas_fee_cap: Option<TokenAmount>,
+    pub gas_premium: Option<TokenAmount>,
+    pub cid: Option<Cid>,
+    pub version: Option<u16>,
+    pub max_fee: Option<TokenAmount>,
+}
+
+impl MpoolPushMessage {
+    pub fn new(to: Address, from: Option<Address>, method: MethodNum, params: Vec<u8>) -> Self {
+        MpoolPushMessage {
+            to,
+            from,
+            method,
+            params,
+            value: TokenAmount::from_atto(0),
+            nonce: None,
+            gas_limit: None,
+            gas_fee_cap: None,
+            gas_premium: None,
+            cid: None,
+            version: None,
+            max_fee: None,
+        }
+    }
+}
+
+impl From<CIDMap> for Option<Cid> {
+    fn from(m: CIDMap) -> Self {
+        m.cid
+            .map(|cid| Cid::from_str(&cid).expect("invalid cid str"))
+    }
+}
+
+impl From<Option<Cid>> for CIDMap {
+    fn from(c: Option<Cid>) -> Self {
+        c.map(|cid| CIDMap {
+            cid: Some(cid.to_string()),
+        })
+        .unwrap_or(CIDMap { cid: None })
+    }
+}
+
+impl From<Cid> for CIDMap {
+    fn from(c: Cid) -> Self {
+        CIDMap {
+            cid: Some(c.to_string()),
+        }
+    }
+}
```

Comment:
These functions shouldn't be conflated into the same file as the message declarations. Please move them to a more appropriate place.

Reply:
what?